File: Sources/DesktopVoiceInput/Services/SmartPostProcessor.swift
import Foundation
import os

@MainActor
final class SmartPostProcessor {
    private static let logger = Logger(subsystem: "com.end.DesktopVoiceInput", category: "SmartPostProcessor")

    private let settings: AppSettings
    private let llmClient: LLMClient
    private let fallback: TranscriptPostProcessor
    private let hotwordStore: HotwordStore

    init(settings: AppSettings, hotwordStore: HotwordStore, llmClient: LLMClient = LLMClient()) {
        self.settings = settings
        self.hotwordStore = hotwordStore
        self.llmClient = llmClient
        self.fallback = TranscriptPostProcessor()
    }

    func processRulesOnly(text: String) -> String {
        var result = fallback.finalize(text)
        result = applyHotwordCorrection(to: result)
        return applyFinalPunctuationRules(to: result)
    }

    func process(text: String, targetApp: String?, targetBundleID: String?) async -> String {
        var result = fallback.finalize(text)

        // 热词模糊替换（规则层，始终生效）
        result = applyHotwordCorrection(to: result)

        if result.isEmpty {
            return result
        }

        // LLM 后处理需要开关打开
        guard settings.postProcessingEnabled else {
            return applyFinalPunctuationRules(to: result)
        }

        let pipeline = resolvePipeline(targetApp: targetApp, targetBundleID: targetBundleID)

        // 执行预设的规则层（如果有）
        if !pipeline.rules.isEmpty {
            result = PostProcessingConfig.applyRules(pipeline.rules, to: result)
        }

        // 执行 LLM 层
        if let prompt = pipeline.llmPrompt, !prompt.isEmpty {
            let config = settings.llmProviderConfig
            guard config.isConfigured else {
                Self.logger.info("LLM not configured, skipping LLM post-processing")
                return result
            }

            let systemPrompt = buildSystemPrompt(basePrompt: prompt)

            do {
                let llmResult = try await llmClient.complete(system: systemPrompt, user: result, config: config)
                if !llmResult.isEmpty {
                    result = llmResult
                }
            } catch {
                Self.logger.error("LLM post-processing failed, using rule-only result. error=\(error.localizedDescription, privacy: .public)")
            }
        }

        // 标点选项是最终输出格式，必须在 LLM 之后再次收口，避免模型重新补回句号。
        result = applyHotwordCorrection(to: result)
        return applyFinalPunctuationRules(to: result)
    }

    private func resolvePipeline(targetApp: String?, targetBundleID: String?) -> PostProcessingConfig.ProcessingPipeline {
        if let prompt = settings.activePostProcessingPrompt {
            return PostProcessingConfig.ProcessingPipeline(
                rules: [.collapseWhitespace, .trimWhitespace],
                llmPrompt: prompt
            )
        }

        return PostProcessingConfig.ProcessingPipeline(
            rules: [.collapseWhitespace, .trimWhitespace]
        )
    }

    private func buildSystemPrompt(basePrompt: String) -> String {
        let replacements = hotwordStore.replacements
        guard !replacements.isEmpty else { return basePrompt }
        let wordList = replacements.map(\.to).joined(separator: "、")
        return basePrompt + "\n\n参考热词表（如果识别结果中有发音相近但拼写不同的词，优先使用热词表中的正确写法）：\(wordList)"
    }

    private func applyHotwordCorrection(to text: String) -> String {
        hotwordStore.applyReplacements(to: text)
    }

    private func applyFinalPunctuationRules(to text: String) -> String {
        let normalized = fallback.finalize(text)
        let punctuationRules = settings.punctuationMode.rules
        return PostProcessingConfig.applyRules(punctuationRules, to: normalized)
    }
}

File: windows/src/GuGuTalk.Core/Services/SmartPostProcessor.cs
using GuGuTalk.Core.Models;
using GuGuTalk.Core.Settings;
using Serilog;

namespace GuGuTalk.Core.Services;

public sealed class SmartPostProcessor
{
    private static readonly ILogger Logger = Log.ForContext<SmartPostProcessor>();

    private readonly AppSettings _settings;
    private readonly HotwordStore _hotwordStore;
    private readonly LLMClient _llmClient;
    private readonly TranscriptPostProcessor _postProcessor = new();

    public SmartPostProcessor(AppSettings settings, HotwordStore hotwordStore, LLMClient llmClient)
    {
        _settings = settings;
        _hotwordStore = hotwordStore;
        _llmClient = llmClient;
    }

    public string ProcessRulesOnly(string text)
    {
        string result = _postProcessor.Finalize(text);
        result = _hotwordStore.ApplyReplacements(result);
        return ApplyPunctuationRules(result);
    }

    public async Task<string> ProcessAsync(string text, string? targetApp = null)
    {
        string result = _postProcessor.Finalize(text);
        result = _hotwordStore.ApplyReplacements(result);

        if (string.IsNullOrEmpty(result)) return result;

        if (!_settings.PostProcessingEnabled)
            return ApplyPunctuationRules(result);

        string? prompt = _settings.ActivePostProcessingPrompt;
        if (string.IsNullOrEmpty(prompt))
            return ApplyPunctuationRules(result);

        if (!_settings.LlmProviderConfig.IsConfigured)
        {
            Logger.Information("LLM not configured, skipping post-processing");
            return ApplyPunctuationRules(result);
        }

        string systemPrompt = BuildSystemPrompt(prompt);

        try
        {
            string llmResult = await _llmClient.CompleteAsync(systemPrompt, result, _settings.LlmProviderConfig);
            if (!string.IsNullOrEmpty(llmResult))
            {
                result = llmResult;
            }
        }
        catch (Exception ex)
        {
            Logger.Error(ex, "LLM post-processing failed, using rule-only result");
        }

        result = _hotwordStore.ApplyReplacements(result);
        return ApplyPunctuationRules(result);
    }

    private string BuildSystemPrompt(string basePrompt)
    {
        if (_hotwordStore.IsEmpty) return basePrompt;

        var wordList = string.Join("、", _hotwordStore.Replacements.Select(r => r.Replacement));
        return basePrompt + $"\n\n参考热词表（如果识别结果中有发音相近但拼写不同的词，优先使用热词表中的正确写法）：{wordList}";
    }

    private string ApplyPunctuationRules(string text)
    {
        string result = _postProcessor.Finalize(text);

        return _settings.PunctuationMode switch
        {
            PunctuationMode.Remove => RemoveAllPunctuation(result),
            PunctuationMode.SpaceReplace => ReplacePunctuationWithSpace(result),
            PunctuationMode.RemoveTrailingPeriod => RemoveTrailingPeriod(result),
            _ => result // Smart: keep as-is
        };
    }

    private static string RemoveAllPunctuation(string text) =>
        new(text.Where(c => !IsCjkPunctuation(c) && !char.IsPunctuation(c)).ToArray()).Trim();

    private static string ReplacePunctuationWithSpace(string text)
    {
        var chars = new List<char>();
        foreach (char c in text)
        {
            if (IsCjkPunctuation(c) || char.IsPunctuation(c))
                chars.Add(' ');
            else
                chars.Add(c);
        }
        return TranscriptPostProcessor.NormalizeSpeechText(new string(chars.ToArray()));
    }

    private static string RemoveTrailingPeriod(string text)
    {
        string trimmed = text.TrimEnd();
        if (trimmed.EndsWith('。') || trimmed.EndsWith('.'))
            return trimmed[..^1];
        return trimmed;
    }

    private static bool IsCjkPunctuation(char c) =>
        c is '，' or '。' or '！' or '？' or '；' or '：' or '、'
        or '（' or '）' or '《' or '》' or '「' or '」' or '『' or '』' or '【' or '】';
}
